### PR TITLE
build(all): use typescript project references

### PR DESCRIPTION
**Pull request by unstubbable**
~This also enables `skipLibCheck` which reduces the compile time significantly. It is also a [recommended compiler option](https://www.typescriptlang.org/tsconfig#skipLibCheck).~

~Note: Explicit usage of third-party code is still checked.~

- [x] switch some scripts in `packages/demos` to js to avoid using `ts-node` (no support for project references)
- [x] fix TypeDoc integration
- [x] investigate change in size-limit result
  The increase of `@feature-hub/core` by 2 kb is caused by a newer version of `semver` (required by `ts-loader`) which introduced memoization to [improve the performance of range parsing](https://github.com/npm/node-semver/commit/e08d9167937e09e8e6fe23aacaf17f892a1d69e1). With this `lru-cache` was added as a dependency which unfortunately leads to the larger overall size of `@feature-hub/core`.
- [x] investigate why Netlify can not build the API docs

### CODE REVIEW

**Comment by unstubbable** [comment]
The necessary changes in `packages/demos` to avoid using `ts-node` (no support for project references) are best done after #631 is merged.

**Comment by fahrradflucht** [comment]
I don't understand why they recommend this option - isn't it just unsafe? You don't know anymore for example of your dep references browser APIs that aren't in your matrix or anything like that.

**Comment by unstubbable** [comment]
The code you are using is type checked. So if a browser API in these code paths is used that you don't support the compiler should be able to catch that (need to test that in a small repro). Of course this assumes that the lib itself would be free of side effects.

**Comment by fahrradflucht** [comment]
💡 I didn't know that. So what you are saying is this?
```ts
// node_modules/a/index.ts
function foo(): string {
  return new Intl.PluralRules().select(0);
}

function bar(): string {
  return "other";

// src/works.ts
import { bar } from "a";

console.log(bar());

// src/broken.ts
import { foo } from "a";

console.log(foo());
```

**Comment by unstubbable** [comment]
@fahrradflucht It's a good thing you started this discussion. Based on your example I created this test repo: https://github.com/unstubbable/skip-lib-check-test Quoting the README from there:

> The documentation says:
> 
> > TypeScript will type check the code you specifically refer to in your app’s source code.
> 
> What it does not mention, but should not be surprising now that I think about it, is that any library code that can not be compiled, but your app's source code refers to, will be regarded as any. Even with noImplicitAny set to true this does not yield a compile error.

This brings me to the conclusion that this compiler option should only be used a stop-gap measure (if at all) for a short period of time, when the types of two libraries (in this case `webpack` and `@types/webpack-dev-middleware`) are incompatible with each other.

**Comment by unstubbable** [comment]
Removed `skipLibCheck` again in 8f57c95. There is currently no issue without `skipLibCheck`, until we decide to update to webpack 5, and https://github.com/DefinitelyTyped/DefinitelyTyped/issues/50773 is not resolved till then.

**Comment by unstubbable** [comment]
Interesting, with `skipLibCheck` removed, TypeDoc fails to generate the API docs with `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory` somewhere in `node_modules/typescript/lib/typescript.js`. 🙄

File: packages/demos/src/node-integrator.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// @ts-check
+
+/**
+ * @param {string} source
+ * @return {any}
```
Comment (clebert) [inline packages/demos/src/node-integrator.js:5]: Use `{{readonly default: import('./app-renderer').AppRenderer}}` and assert the type internally with e.g. `typeof mod.exports.default === 'function'`

File: tsconfig.settings.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (clebert) [inline tsconfig.settings.json:1]: Is it possible to also activate `noUncheckedIndexedAccess`?

**Review by clebert — CHANGES_REQUESTED** [review]

File: tsconfig.settings.json (hunk 2 above)
Comment (unstubbable) [reply tsconfig.settings.json:1]: Yes, we should definitely enable that. #643

**Review by clebert — DISMISSED** [review]

**Review by clebert — APPROVED** [review]

**Pull request merged**